libraries/chain/webassembly/runtimes/eos-vm-oc/code_cache.cpp

```diff
@@ -107,6 +107,8 @@ std::tuple<size_t, size_t> code_cache_async::consume_compile_thread_queue() {
 }
 
 const code_descriptor* const code_cache_async::get_descriptor_for_code(const digest_type& code_id, const uint8_t& vm_version) {
+   std::lock_guard<std::shared_mutex> g(get_descriptor_async_mutex);
```

Comment:
I believe this lock is not sufficient. Since all executors share the same code_cache, that means you cannot evict an entry in the code cache that is currently running in any executor. While the lock protects the integrity of the data structures internal to code_cache, if any incoming code is handled,
https://github.com/AntelopeIO/leap/blob/c2ef8b2f598dfe8e14e76cbadf0f5d70db46ea4a/libraries/chain/webassembly/runtimes/eos-vm-oc/code_cache.cpp#L116-L117
there is a possibility of running an eviction,
https://github.com/AntelopeIO/leap/blob/c2ef8b2f598dfe8e14e76cbadf0f5d70db46ea4a/libraries/chain/webassembly/runtimes/eos-vm-oc/code_cache.cpp#L384-L387
which could potentially remove entries in the code cache that other executors are currently running.

As a quick fix, I would suggest `get_descriptor_for_code()` having two behaviors: on the main thread (during single threaded write window) it behaves how it does today where it services outstanding compiles & considers eviction. But then during the parallel read only window it only returns existing entries in the code_cache (it could be done w/o a mutex even, I believe).

Reply:
Thanks! 

Do you mean we should skip the `if` block https://github.com/AntelopeIO/leap/blob/c2ef8b2f598dfe8e14e76cbadf0f5d70db46ea4a/libraries/chain/webassembly/runtimes/eos-vm-oc/code_cache.cpp#L113 for read only threads and run the remainder code for read only threads, and on main thread we keep the existing behavior?

Reply:
Right, if not the main thread skip the entirety of that `if`. But also, after this,
https://github.com/AntelopeIO/leap/blob/c2ef8b2f598dfe8e14e76cbadf0f5d70db46ea4a/libraries/chain/webassembly/runtimes/eos-vm-oc/code_cache.cpp#L137-L141
if not main thread, simply return `nullptr` (don't continue on to blacklist processing and all the other stuff).

I see now
https://github.com/AntelopeIO/leap/blob/c2ef8b2f598dfe8e14e76cbadf0f5d70db46ea4a/libraries/chain/webassembly/runtimes/eos-vm-oc/code_cache.cpp#L139
and since that modifies the container you would need either a mutex or to skip this when non-main-thread too. Usually I lean toward avoiding mutexes, and I do kind of feel like read only trxs shouldn't affect the LRU policy, so I'd probably suggest avoiding this to ensure it can be mutex free. But you can decide what you want to do in that regard.

Reply:
Thanks, I will choose no mutex option.

I added a mutex to `code_cache_sync::get_descriptor_for_code_sync`. Is it not necessary? Should I remove it too or do the similar to `code_cache_async`?

Reply:
I have made a similar change to `code_cache_sync`, removing the mutex.